PR to collaborate on IR

Jake2206 opened the pull request:
- [x] Added types to LHS of matrix/array decl and add types to ir gen file
- [x] 🐛 Fix bug with fdecl.sformals type
- [x] Update Translate Function
- [x] Add ability for global vars
- [x] Function declarations
- [x] Update build_expr function (add arrays and matrices)
- [x] Implement lambda functions
- [x]  Fix print function bug (can't initialize arrays, matrices, lambda functions, function calls, binary ops)
- [x] Arrays of floats
- [x] Matrix
- [ ] Print functions for non-ints
- [x] Array and matrix pointer bugs (I think the pointer types are not correctly configured but I haven't tested much)
- [x] Implement binary operations for non-floats

TO DO:
1. Hooking C functions (or functions available through LLVM). Might make our lives easier when it comes to doing actual computation.
   An example:

let printf_t : L.lltype =
L.var_arg_function_type i32_t [| L.pointer_type i8_t |] in
let printf_func : L.llvalue =
L.declare_function "printf" printf_t the_module

links the printf stdlib function to the program, as far as I can tell. It is invoked in build_expr with  
 .....
| SCall ("print", [e]) ->
L.build_call printf_func [| int_format_str ; (build_expr builder e) |]
"printf" builder

If we want to use other functionality -- we probably have to use this format or something similar.

Jake2206 committed:
Add types to LHS of matrix/array decl and add types to ir gen file

Etesam913 committed:
Merge branch 'IR_gen' of https://github.com/Jake2206/Comma into IR_gen

Jake2206 committed:
Merge branch 'main' into IR_gen

Etesam913 committed:
🐛 Testing bug

Etesam913 committed:
Merge branch 'IR_gen' of https://github.com/Jake2206/Comma into IR_gen

Etesam913 committed:
🔥 Removed extra param in ast string_of_program func

Etesam913 committed:
✅ Added passing tests for global var decl

Etesam913 commented:
It seems that we are dropping the ability to have floating statements as that was a miscommunication. This is how the parser is handling it currently, you can see two arrays:
<img width="778" alt="Screen Shot 2022-04-27 at 4 33 43 PM" src="https://user-images.githubusercontent.com/55665282/165626234-cf35b2b4-70f9-4e59-923b-740a162fdde7.png">

This means that the SAST.sprogram can be made pretty much identical to the one in microc. Therefore, our IR generation will be much more similar to microc, so hopefully there will be less work to do.

Etesam913 committed:
✨Globals should work

Just reimplemented the globals from microc.

🐛 There is a weird syntax bug with our bind definition.

Jake2206 committed:
Fix bug with fdecl.sformals type

Jake2206 committed:
Implement lambda functions, add basic expr types (not array or matrix)

Jake2206 committed:
Arrays and matrices implemented, IRgen now compiles

Jake2206 committed:
First test compiled w/ no error

Jake2206 committed:
Fixed/updated for loops in irgen.ml

Jake2206 committed:
Assignbind mid-update bug fix

Commit:
Constructed standard library in C outline, implemented parseCSV

Etesam913 commented:
I think I have come close to figuring out why printing variables doesn't work. More specifically printing a variable that is initialized and assigned a value in one line. If the variable is initialized, and then assigned later on, it will print out fine. Here is an image to understand what I wrote above:
<img width="595" alt="Screen Shot 2022-05-06 at 1 28 06 AM" src="https://user-images.githubusercontent.com/55665282/167072656-c3c69455-67a0-43d9-980f-f2829c2f65e8.png">

I suspect the reason why this is happening is due to the SAssign type.
<img width="628" alt="Screen Shot 2022-05-06 at 1 30 03 AM" src="https://user-images.githubusercontent.com/55665282/167072746-597e58ee-297c-4a2c-9443-d6f76515af70.png">
The SAssign case is where the value gets loaded into the map from my understanding. It seems like a statement like 
`int a;` triggers SAssign while a statement like `int a = 4;` does not trigger SAssign.
 
Looking at the sast.ml file, I think this becomes more clear.
<img width="391" alt="Screen Shot 2022-05-06 at 1 42 09 AM" src="https://user-images.githubusercontent.com/55665282/167073858-7e11ae5f-dee8-456d-9885-a30755c3d489.png">


For SAssign to be called the definition would have to be changed to `typ * string * sexpr`. However, this would reduce support for statements like `a=5`. We probably want to support both so we may have to define a new SAST and AST case to catch when declaration and initialization happen in one line.

Jake2206 committed:
Basic global variable assignments implemented

Jake2206 commented:
> It seems that we are dropping the ability to have floating statements as that was a miscommunication. This is how the parser is handling it currently, you can see two arrays: <img alt="Screen Shot 2022-04-27 at 4 33 43 PM" width="778" src="https://user-images.githubusercontent.com/55665282/165626234-cf35b2b4-70f9-4e59-923b-740a162fdde7.png">
> 
> This means that the SAST.sprogram can be made pretty much identical to the one in microc. Therefore, our IR generation will be much more similar to microc, so hopefully there will be less work to do.

Yes, you are correct. I've been working on fixing this issue but we will need to define what we want to be able to do with global variables. The big hurdle on this is that we don't have access to build_expr when we initialize global variables and it doesn't seem easy to extricate that function out of build_function to put it in scope for initializing the globals. I implemented basic global variable assignments (ints, bools, chars, doubles), but we will need to pretty much recreate the build_expr function to be able to use any expression as a global variable unless we can figure out how to get around the scoping issue.

edit: sorry I replied to the wrong comment, I meant o reply to the one about the global var bug. I think I just pushed a fix though.

Jake2206 committed:
Fixed global variable assignment bug

Jake2206 committed:
Fix lambda function bug

Commit:
Standard library function implementation. Code compiles, but C code does not execute

Etesam913 committed:
✨ Implemented retrieveElement method in stdlib.c

♻️  Changed elements property in matrix struct from double * to double ** for easier matrix operations down the road

⚙️ Updated parseCSV function to work with elements being double **

Etesam913 committed:
✨ Implemented scalarMulti & scalarDiv methods

Removed free statements related to elements in parseCSV as freeing the values at that point leads to undesired segmentation faults

Jake2206 committed:
Array/matrix pointer bug fix, S/R conflict from lambda decl fix, update to matrix syntax, README update, update to lambda syntax, array float types

Etesam913 committed:
✨ Implemented addMatrix & subtractMatrix methods

💾 Added a deallocateMatrix method to prevent memory leaks.

Jake2206 committed:
Remove higher order function partial implementation (too difficult to finish with remaining time)

Commit:
Finished implementing C standard library code

Commit:
Merging

Commit:
Finish merging

Jake2206 committed:
Update/add printing formatters

Jake2206 committed:
Merge branch 'IR_gen' into StdLibBranch

Commit:
Fixed libstd.c

lisi-w committed:
added script to automate testing of comma text files

Jake2206 committed:
Print function handling of ints, chars, bools, doubles, and char arrays. Matrix llvm type to double**

Jake2206 committed:
Clean up comments, add comments

lisi-w committed:
updated script

lisi-w committed:
add make clean

Jake2206 committed:
Remove references to eif, too hard too implement w/ remaining time

ecs2220 committed:
Merge pull request #15 from Jake2206/Test-Script

Test script

Commit:
Print Matrix implemented

Commit:
Saving irgen.ml

Commit:
Merging with local branch

Commit:
C LIBRARY WORKS NOW HALLELUJIAH

Jake2206 committed:
Implement modulo, add tests (gcd example, for loop with nested lambda functions to increment))

Jake2206 committed:
Update readme, fix small error in tests.sh (still not working for me though)

Jake2206 committed:
Small optimization on global string pointers, clean up commented out old/unused code

Jake2206 committed:
test script update, global doubles,  modulo precedence fix

Jake2206 committed:
Merge pull request #14 from Jake2206/StdLibBranch

Std lib branch
